test/src/box/DedicatedWorkerRegistry.js

```diff
@@ -0,0 +1,256 @@
+    LogBox = foam.lookup('foam.box.LogBox');
+  });
+
+  function ctxFactory() {
+    var ctx = Context.create();
+    var mockRegistry = MockRegistry.create(null, ctx);
+    ctx.registry = Registry.create({
+      delegate: mockRegistry,
+      getDedicatedWorkerKey: function(box) {
+        return box.serviceName; // Only for the purpose of these tests.
+      }
+    });
+    return ctx;
+  }
+
+  it('should forward all registrations to delegate by default', function() {
+    var ctx = ctxFactory();
+    var mockRegistry = ctx.registry.delegate;
+
+    ctx.registry.register(null, null, foam.box.LogBox.create());
+    ctx.registry.register('LogBox2', null, foam.box.LogBox.create());
+
+    expect(Object.keys(mockRegistry.registry).length).toBe(2);
+    expect(mockRegistry.actions[0].action).toBe('register');
+    expect(foam.box.LogBox.isInstance(mockRegistry.actions[0].box)).toBe(true);
+    expect(mockRegistry.actions[1].action).toBe('register');
+    expect(foam.box.LogBox.isInstance(mockRegistry.actions[1].box)).toBe(true);
+  });
+
+  it('should forward register requests of unknown services to delegate', function() {
+    var ctx = ctxFactory();
+    var mockRegistry = ctx.registry.delegate;
+
+    var registryCount = function() {
+      return Object.keys(ctx.registry.localRegistry.registry).length;
+    }
+
+    var dedicated = foam.box.LogBox.create();
+    dedicated.serviceName = 'Box';
+    var nonDedicated = foam.box.LogBox.create();
+
+    ctx.registry.register(null, null, dedicated);
+    // One register for ForkFox, one for dedicatedWorkers_
+    expect(registryCount()).toBe(2);
+    expect(mockRegistry.actions.length).toBe(0); // Delegate should not be hit
+
+    ctx.registry.register('TestName', null, dedicated);
+    // No reregister for ForkBox, one for dedicatedWorkers_
+    expect(registryCount()).toBe(3);
+    expect(mockRegistry.actions.length).toBe(0);
+
+    ctx.registry.register(null, null, nonDedicated);
+    // Verify dedicated registery was not hit
+    expect(registryCount()).toBe(3);
+    expect(mockRegistry.actions.length).toBe(1);
+
+    ctx.registry.register('Test2', null, nonDedicated);
+    // Verify dedicated registry was not hit
+    expect(registryCount()).toBe(3);
+    expect(mockRegistry.actions.length).toBe(2);
+  });
+
+  it('should send messages with services registered with dedicated registry', function(done) {
+    var ctx = ctxFactory();
+    var mockRegistry = ctx.registry.delegate;
+    var dao = foam.dao.ArrayDAO.create({ of: 'Object' });
+    var box = foam.box.SkeletonBox.create({ data: dao });
+    box.serviceName = "ArrayDAO"; // Enabling dedicated service worker.
+
+    // Perform remote registration of service.
+    var register = ctx.registry.register(null, null, box);
+
+    // Creating a stub on local for accessing method.
+    var stub = foam.core.StubFactorySingleton.create().get(foam.dao.DAO)
+        .create({ delegate: register }, ctx.registry.localRegistry);
```

Comment:
Should the user need to know which registry the box above was registered in? If `ctx` is provided instead (i.e. allowing `DedicatedWorkerRegistry` to handle the registration), it will perform registrations for the stub in the delegate's registry. This results in issues when we call `stub.put(obj)` below (it will not be able to find the ReplyBox).

A stack trace captured at the lookup failure:
```
code | @ | BoxRegistryBox.js:48
send | @ | DedicatedWorkerRegistry.js:146
send | @ | NamedBox.js:53
code | @ | SubBox.js:42
(anonymous function) | @ | Promised.js:56
_tickCallback | @ | internal/process/next_tick.js:109
_Promise.resolve (async)_
(anonymous function) | @ | ForkBox.js:81
send | @ | ReplyBox.js:42
code | @ | BoxRegistryBox.js:51
(anonymous function) | @ | net.js:471
```

To be clear: providing `ctx.registry.localRegistry` will work fine. I'm just not sure if the user should be required to specify the correct registry.